tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1264 @@
+    ABSL_LOG(INFO) << "Using experimental_s3_rate_limiter";
+  }
+  return driver;
+}
+
+/// A ReadTask is a function object used to satisfy a
+/// S3KeyValueStore::Read request.
+struct ReadTask : public RateLimiterNode,
+                  public internal::AtomicReferenceCount<ReadTask> {
+  IntrusivePtr<S3KeyValueStore> owner;
+  std::string endpoint;
+  kvstore::ReadOptions options;
+  Promise<kvstore::ReadResult> promise;
+
+  int attempt_ = 0;
+  absl::Time start_time_;
+
+  ReadTask(IntrusivePtr<S3KeyValueStore> owner, std::string endpoint,
+           kvstore::ReadOptions options, Promise<kvstore::ReadResult> promise)
+      : owner(std::move(owner)),
+        endpoint(std::move(endpoint)),
+        options(std::move(options)),
+        promise(std::move(promise)) {}
+
+  ~ReadTask() { owner->admission_queue().Finish(this); }
+
+  static void Start(void* task) {
+    auto* self = reinterpret_cast<ReadTask*>(task);
+    self->owner->read_rate_limiter().Finish(self);
+    self->owner->admission_queue().Admit(self, &ReadTask::Admit);
+  }
+
+  static void Admit(void* task) {
+    auto* self = reinterpret_cast<ReadTask*>(task);
+    self->owner->executor()(
+        [state = IntrusivePtr<ReadTask>(self, internal::adopt_object_ref)] {
+          state->Retry();
+        });
+  }
+
+  void Retry() {
+    if (!promise.result_needed()) {
+      return;
+    }
+
+    // TODO(sjperkins). Introduce S3 versioning logic here
+    /// Reads contents of a S3 object.
+    // std::string media_url = tensorstore::StrCat(resource, "?alt=media");
+
+    // Add the ifGenerationNotMatch condition.
+    // AddGenerationParam(&media_url, true, "ifGenerationNotMatch",
+    //                    options.if_not_equal);
+    // AddGenerationParam(&media_url, true, "ifGenerationMatch", options.if_equal);
+
+    // Assume that if the user_project field is set, that we want to provide
+    // it on the uri for a requestor pays bucket.
+    // AddUserProjectParam(&media_url, true, owner->encoded_user_project());
+
+    // AddUniqueQueryParameterToDisableCaching(media_url);
+
+    // TODO: Configure timeouts.
+    auto maybe_credentials = owner->GetCredentials();
+    if (!maybe_credentials.ok()) {
+      promise.SetResult(maybe_credentials.status());
+      return;
+    }
+
+    S3RequestBuilder request_builder("GET", endpoint);
+    S3Credentials credentials;
+
+    if (maybe_credentials.value().has_value()) {
+      credentials = std::move(*maybe_credentials.value());
+    }
+
+    start_time_ = absl::Now();
+    bool result;
+
+    request_builder.AddRangeHeader(options.byte_range, result);
+    auto request = request_builder
+            .EnableAcceptEncoding()
+            .AddHeader(absl::StrCat("host: ", std::string(internal::ParseGenericUri(owner->endpoint_).authority_and_path)))
+            .AddHeader(absl::StrCat("x-amz-content-sha256: ", kEmptySha256))
+            .AddHeader(absl::FormatTime("x-amz-date: %Y%m%dT%H%M%SZ", start_time_, absl::UTCTimeZone()))
```

Comment:
> I've implemented basic S3 credentials provider for the environment variables and credentials file (but not EC2 Metadata).
> 
> Also, AWS bucket region inspection functionality to construct a virtual path style endpoint as this gets the endpoint correct for the LocationConstraint.
> 
> `Read`, `Write`, `List` and `Delete` functionality was cut-n-pasted over from the GVS KVstore without much thought -- I'm unsure as to whether the same response codes will still apply.
> 
> @laramiel I still need to incorporate your object metadata review comments.

The read operation will probably be pretty similar.  S3 doesn't seem to document all of the response codes for GetObject (https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html) but probably the GCS ones would apply.

There are some key differences between the S3 API and GCS:

- TensorStore needs a `StorageGeneration` that uniquely identifies the object content.  GCS has a `generation` field that serves this purpose directly; it is essentially a globally-consistent timestamp at which the object was written.  S3 does not have a `generation` metadata field, but instead it has: `ETag` and `LastModified`.  `ETag` is supposed to serve as a unique identifier, but on S3 it is sometimes an md5 hash for which there are collisions.  I think it would be sufficient to use `LastModified` as the generation.  Alternatively both LastModified and ETag could be combined to produce the storage generation.
- GCS supports conditional writes, which are used for Write operations where `if_equal` is specified.  S3 does not support this; the best we can do is issue an extra `HeadObject` request before writing if `if_equal` is specified in order to check that the generation matches (this is inherently racy if there are concurrent writes by other machines).  Users simply have to be aware that they cannot rely on `if_equal` being race-free with S3.
- GCS returns `List` responses in a different format; as we discussed previously, the S3 list response is returned in XML format rather than a JSON format.

Reply:
> Maybe `S3RequestBuilder` could have:
> 
> * `AddHost`
> * `AddPayloadHash`
> * `AddDate`
> 
> methods, or even
> 
> * `AddRequiredHeaders(host, payload_hash, date)`
> 
> which would avoid passing in `payload_hash` + `start_time_` twice.
> 
> I experimented with the first approach but this produced some conflicts with the AWS4 Signature request test cases, so I reverted to the explicit approach here.

Is there a reason why `BuildRequest` can't just add the necessary headers, since you are already passing that information to it?

Reply:
> Is there a reason why BuildRequest can't just add the necessary headers, since you are already passing that information to it?

Yes, there's probably some sensible refactoring possible here. I've added this to the TODO list in the first comment for later polishing as this is working with AWS S3, as is. Is that OK?